File: java/com/lc/lc189.java
package lc;

public class lc189 {
    public void rotate(int[] nums, int k) {
        k = k % nums.length;
        reverse(nums, 0, nums.length);
        reverse(nums, 0, k);
        reverse(nums, k, nums.length);
    }

    public void reverse(int[] nums, int l, int r) {
        while (l < r) {
            int temp = nums[l];
            nums[l] = nums[r - 1];
            nums[r - 1] = temp;
            l++;
            r--;
        }
    }
}

File: go/lc189.go
package main

func rotate(nums []int, k int) {
	k = k % (len(nums))

	swap := func(i, j int) {
		for i < j {
			t := nums[i]
			nums[i] = nums[j]
			nums[j] = t
			i++
			j--
		}
	}

	swap(0, len(nums)-1)
	swap(0, k-1)
	swap(k, len(nums)-1)
}
